File: 1625-lexicographically-smallest-string-after-applying-operations/1625-lexicographically-smallest-string-after-applying-operations.rb
# @param {String} s
# @param {Integer} a
# @param {Integer} b
# @return {String}
def find_lex_smallest_string(s, a, b)
  seen = {}
  queue = [s]
  smallest = s

  while queue.any?
    current = queue.shift
    next if seen[current]
    seen[current] = true
    smallest = [smallest, current].min

    # Apply operation 1: Add `a` to all odd-indexed positions
    modified = current.chars.map.with_index do |chr, i|
      i.odd? ? (chr.to_i + a) % 10 : chr.to_i
    end.join
    queue.push(modified) unless seen[modified]

    # Apply operation 2: Rotate string by `b` positions
    rotated = current.chars.rotate(-b).join
    queue.push(rotated) unless seen[rotated]
  end

  smallest
end

File: 1625-lexicographically-smallest-string-after-applying-operations/1625-lexicographically-smallest-string-after-applying-operations.py
class Solution:
    def findLexSmallestString(self, s: str, a: int, b: int) -> str:
        seen = set()
        queue = deque([s])
        smallest = s

        while queue:
            current = queue.popleft()
            if current in seen:
                continue
            seen.add(current)
            smallest = min(smallest, current)

            # Apply operation 1: Add `a` to all odd-indexed positions
            modified = list(current)
            for i in range(1, len(modified), 2):
                modified[i] = str((int(modified[i]) + a) % 10)
            modified = ''.join(modified)
            if modified not in seen:
                queue.append(modified)

            # Apply operation 2: Rotate string by `b` positions
            rotated = current[-b:] + current[:-b]
            if rotated not in seen:
                queue.append(rotated)

        return smallest        